third_party/move/move-vm/runtime/src/loader/function.rs

```diff
@@ -60,16 +72,256 @@ pub struct LoadedFunction {
+                fun,
+                fun_inst,
+                mask,
+            },
+        ))))
+    }
+
+    pub(crate) fn expect_this_impl(
+        fun: &dyn AbstractFunction,
+    ) -> PartialVMResult<&LazyLoadedFunction> {
+        fun.downcast_ref::<LazyLoadedFunction>().ok_or_else(|| {
+            PartialVMError::new_invariant_violation("unexpected abstract function implementation")
+        })
+    }
+
+    /// Access name components independent of resolution state. Since RefCell is in the play,
+    /// the accessor is passed in as a function.
+    pub(crate) fn with_name_and_inst<T>(
+        &self,
+        action: impl FnOnce(Option<&ModuleId>, &IdentStr, &[TypeTag]) -> T,
+    ) -> T {
+        match &*self.0.borrow() {
+            LazyLoadedFunctionState::Unresolved {
+                data:
+                    SerializedFunctionData {
+                        module_id,
+                        fun_id,
+                        fun_inst,
+                        ..
+                    },
+                ..
+            } => action(Some(module_id), fun_id, fun_inst),
+            LazyLoadedFunctionState::Resolved { fun, fun_inst, .. } => {
+                action(fun.module_id(), fun.name_id(), fun_inst)
+            },
+        }
+    }
+
+    /// Executed an action with the resolved loaded function. If the function hasn't been
+    /// loaded yet, it will be loaded now.
+    #[allow(unused)]
+    pub(crate) fn with_resolved_function<T>(
+        &self,
+        storage: &dyn ModuleStorage,
+        action: impl FnOnce(Rc<LoadedFunction>) -> PartialVMResult<T>,
+    ) -> PartialVMResult<T> {
+        let mut state = self.0.borrow_mut();
+        match &mut *state {
+            LazyLoadedFunctionState::Resolved { fun, .. } => action(fun.clone()),
+            LazyLoadedFunctionState::Unresolved {
+                resolution_error: Some(e),
+                ..
+            } => Err(e.clone()),
+            LazyLoadedFunctionState::Unresolved {
+                data:
+                    SerializedFunctionData {
+                        module_id,
+                        fun_id,
+                        fun_inst,
+                        mask,
+                        captured_layouts,
+                    },
+                resolution_error,
+            } => match Self::resolve(
+                storage,
+                module_id,
+                fun_id,
+                fun_inst,
+                *mask,
+                captured_layouts,
+            ) {
+                Ok(fun) => {
+                    let result = action(fun.clone());
+                    *state = LazyLoadedFunctionState::Resolved {
+                        fun,
+                        fun_inst: fun_inst.clone(),
+                        mask: *mask,
+                    };
+                    result
+                },
+                Err(e) => {
+                    *resolution_error = Some(e.clone());
```

Comment:
Was wondering why the interior mutability is needed here for error. If the resolution failed this function would fail and you don't need to store the error msg here?s

Reply:
resolution_error field is now removed.